### DISCUSSION TITLE
How to add irradiation growth model

### DISCUSSION THREAD

**Discussion by jin0123456** [idaholab/moose#24912] [Q&A Modules: Solid mechanics]
Hello,

I was wondering how to add irradiation growth model to the simulation. And I have a model of the irradiation growth rate. Which material should I use? 
Hope to get some ideas.

**Comment by GiudGiud** [comment]
Hello

I don't think this is available in MOOSE.
You ll want to ask for access to Bison or marmot, depending on the scale you care about
https://ncrcaims.inl.gov/Identity/Account/Register

@recuero 

Guillaume

**Reply by jin0123456** [reply]
![image](https://github.com/idaholab/moose/assets/130671073/649827a3-0432-482f-b853-2517d4368c5d)
Here `ε_glc` is strain in axial direction and `ε_grc` is strain in radial direction. I'm just trying to add a simple strain model that shows up in anisotropy. I wonder how I can add this eigenstrain model using corresponding material like `Compute Thermal Expansion Eigenstrain`?
Thanks.

**Reply by recuero** [reply]
You can inherit from (AD)`ComputeEigenstrainBase` and implement a directional constitutive behavior, which we already have in various places in the MOOSE modules.  If you search for "anisotropy" you'll find a few instances of it. Since you can define a thermal eigenstrain in the form of a second order tensor, there is no limitation as to behavior in the radial or axial directions.

**Reply by jin0123456** [reply]
Thanks for your reply. I wonder if there's an example that could help for my situation cause I didn't find one. Or examples about how can I write the corresponding tensor from this strain model?
Thanks again.

**Reply by recuero** [reply]
If you look at `ComputeCrystalPlasticityEigenstrainBase` in the tensor mechanics module, you'll find that a `RankTwoTensor` eigenstrain is being computed.  You can replace that definition of the eigenstrain with your own definition and include it in your mechanics. The example may not fit exactly what you want to do, but, as I understand what you want to do, all the ingredients are there.

**Reply by jin0123456** [reply]
OK, I will try. Thanks.

**Reply by jin0123456** [reply]
Here I inherited `ComputeEigenstrainBase`  for my irradiation growth model and applied it to the input file. But I found that I did not seem to have succeeded in applying this model to my simulations cause the simulation results did not change after adding this part.
Here is the .h and .C,
```
#pragma once

#include "ComputeEigenstrainBase.h"

#include "RankTwoTensor.h"
#include "Function.h"

class Function;
template <bool is_ad>
class ComputeIrridiationTempl : public ComputeEigenstrainBaseTempl<is_ad>
{
public:
  static InputParameters validParams();

  ComputeIrridiationTempl(const InputParameters & parameters);

protected:

  virtual void computeQpEigenstrain() override;
  virtual void computeQpProperties() override;

  const Function & _neu;
  const GenericMaterialProperty<Real, is_ad> & _prefactor;

  GenericMaterialProperty<Real, is_ad> & _strain_glc;
  GenericMaterialProperty<Real, is_ad> & _strain_grc;
  GenericMaterialProperty<Real, is_ad> & _strain_cir;
  GenericMaterialProperty<RankTwoTensor, is_ad> & _eigen_base_tensor;

  using Material::_qp;
  using Material::_t;
  using Material::_q_point;
  using ComputeEigenstrainBaseTempl<is_ad>::_eigenstrain;
};

typedef ComputeIrridiationTempl<false> ComputeIrridiation;
typedef ComputeIrridiationTempl<true> ADComputeIrridiation;
```
```
#include "ComputeIrridiation.h"

registerMooseObject("WorkerApp", ComputeIrridiation);
registerMooseObject("WorkerApp", ADComputeIrridiation);

template <bool is_ad>
InputParameters
ComputeIrridiationTempl<is_ad>::validParams()
{
  InputParameters params = ComputeEigenstrainBase::validParams();
  params.addClassDescription("Computes a constant Eigenstrain");
  params.addParam<FunctionName>("neu", 0.0, "Neutron");
  params.addParam<MaterialPropertyName>(
      "prefactor", 1.0, "Name of material property defining the variable dependence");
  return params;
}

template <bool is_ad>
ComputeIrridiationTempl<is_ad>::ComputeIrridiationTempl(const InputParameters & parameters)
  : ComputeEigenstrainBaseTempl<is_ad>(parameters),
    _prefactor(this->template getGenericMaterialProperty<Real, is_ad>("prefactor")),
    _strain_glc(this->template declareGenericProperty<Real, is_ad>("strain_glc")),
    _strain_grc(this->template declareGenericProperty<Real, is_ad>("strain_grc")),
    _strain_cir(this->template declareGenericProperty<Real, is_ad>("strain_cir")),
    _eigen_base_tensor(this->template declareGenericProperty<RankTwoTensor, is_ad>("eigen_base_tensor")),
    _neu(this->getFunction("neu"))
{
  
}

template <bool is_ad>
void
ComputeIrridiationTempl<is_ad>::computeQpProperties()
{
  _strain_glc[_qp] = 2.18e-21 * std::pow(_neu.value(_t, _q_point[_qp]) * _t / 10000., 0.845);
  //_strain_glc[_qp] = 0.1;
  _strain_grc[_qp] = std::pow((1. + _strain_glc[_qp]), -0.5) - 1;
  _strain_cir[_qp] = 0.0;
  _eigen_base_tensor[_qp].fillFromInputVector({_strain_grc[_qp],_strain_glc[_qp],_strain_cir[_qp]});
}



template <bool is_ad>
void
ComputeIrridiationTempl<is_ad>::computeQpEigenstrain()
{
  // Define Eigenstrain
  _eigenstrain[_qp] = _eigen_base_tensor[_qp] * _prefactor[_qp];
}
```
Here is part of input,
```
[Modules/TensorMechanics/Master]
  [fuel]
    strain = FINITE
    add_variables = true
    eigenstrain_names = 'thermal_expansion_eigenstrain'
    generate_output = ' stress_zz vonmises_stress'
    block = 'fuel'
  []
  [clad]
    strain = FINITE
    add_variables = true
    eigenstrain_names = 'thermal_expansion_eigenstrain irridiation_growth_eigenstrain'
    generate_output = ' stress_zz vonmises_stress creep_strain_xx'
    block = 'clad'
  []
[]
```
```
  [CZ_irridiation_growth]
    type = ComputeIrridiation
    neu = neu_flux_fun
    eigenstrain_name = irridiation_growth_eigenstrain
    block = 'clad'
  []
```
Where did I go wrong about all this? @recuero

**Reply by jin0123456** [reply]
@GiudGiud any ideas?
